### ISSUE TITLE
Handling comments

### ISSUE THREAD

**Issue by LeRondPoint** [themosis/framework#327]
Is there a Themosis way of handling comments?  Given the restrictive way Wordpress handles comments, I'm not sure if it's possible to have a @comments/@endcomments directive but it's what I would have expected.

Ath the moment, I put my comment markup in a scout partial file but the different wp calls are not working.
Should I revert to using the regular comments template?

**Comment by LeRondPoint** [comment]
Found the following works:
                    {{ comments_template('/resources/views/partials/comments.scout.php') }}

Maybe a simple @comments('partials.comments') would suffice!

I smell a PR

**Comment by jlambe** [comment]
Currently there is no `@comments` directive.
For the moment, comments are automatically retrieved on a single post view if you're using the WordPress loop `@loop` `@endloop`
Actually I use core functions inside a loop:
```php
@loop
    // post code
    @if(comments_open())
            <section id="comments" class="comments">
                    @if(have_comments())
                        {{ comments_template('/resources/views/blog/comments.php') }}
                    @endif
                    <?php comment_form([
                        'title_reply' => __('Réagissez à cet article', THEME_TEXTDOMAIN),
                        'class_submit' => 'submit button button-secondary'
                    ]); ?>
            </section>
    @endif
@endloop
```
Your code is also working. But as `comments_template()` don't know Scout (or Blade or Twig), you can only write vanilla PHP into the `comments.(scout.)php` file. Which contains this usually:
```php
<ol class="comment-list">
    <?php wp_list_comments(); ?>
</ol>
<div class="comments-pagination">
    <?php paginate_comments_links(); ?>
</div>
```
But yes, the `@comments()` directive can be a shortcut as well ;)

**Comment by LeRondPoint** [comment]
I tried inside the loop but it's not working.
I thought that scout would not work with comment_template but actually it does...  which left me scratching my head...

**Proof**

comments.scout.php
`
@unless (post_password_required())
    <div id="comments" class="comments-area">
       @if (have_comments())
            <h3 class="comments-title">
                {{
                    sprintf(_n('One thought on &ldquo;%2$s&rdquo;',
                               '%1$s thoughts on &ldquo;%2$s&rdquo;',
                               get_comments_number(),
                               'mn_flow' ),
                            number_format_i18n( get_comments_number() ),
                            Loop::title() )
                }}
            </h3>

            @include('partials.commentnav', array(
                     'commentNavId' => 'comment-nav-above'))
        
            <ol class="comment-list">
                {{
                    wp_list_comments( array(
                        'style'      => 'ol',
                        'short_ping' => true,
                        'avatar_size'=> 34,
                    ))
                }}
            </ol>
        
            @include('partials.commentnav', array(
                     'commentNavId' => 'comment-nav-below'))
        
            @if (!comments_open())
                <p class="no-comments">{{ _e( 'Comments are closed.', 'mn_flow' ) }}</p>
            @endif
        @endif
    
        <?php 
    /*    $args = array(
          'id_form'           => 'commentform',
          'id_submit'         => 'submit',
          'title_reply'       => __( 'Leave a Reply', 'mn_flow' ),
          'title_reply_to'    => __( 'Leave a Reply to %s', 'mn_flow' ),
          'cancel_reply_link' => __( 'Cancel Reply', 'mn_flow' ),
          'label_submit'      => __( 'Post Comment', 'mn_flow' ),
    
          'comment_field' =>  '<p class="comment-form-comment"><label for="comment">' . __( 'Comment', 'mn_flow' ) .
            '</label><textarea id="comment" class="form-control" name="comment" cols="45" rows="8" aria-required="true">' .
            '</textarea></p>',
    
          'must_log_in' => '<p class="must-log-in">' .
            sprintf(
              __( 'You must be <a href="%s">logged in</a> to post a comment.', 'mn_flow' ),
              wp_login_url( apply_filters( 'the_permalink', get_permalink() ) )
            ) . '</p>',
    
          'logged_in_as' => '<p class="logged-in-as">' .
            sprintf(
            __( 'Logged in as <a href="%1$s">%2$s</a>. <a href="%3$s" title="Log out of this account">Log out?</a>' ),
              admin_url( 'profile.php' ),
              $user_identity,
              wp_logout_url( apply_filters( 'the_permalink', get_permalink( ) ) )
            ) . '</p>',
    
          'comment_notes_before' => '<p class="comment-notes">' .
            __( 'Your email address will not be published.', 'mn_flow' ) . '</p>',
    
          'comment_notes_after' => '',
    
          'fields' => apply_filters( 'comment_form_default_fields', array(
    
            'author' =>
              '<p class="comment-form-author">' .
              '<label for="author">' . __( 'Name', 'mn_flow' ) . '</label> ' .
              ( $req ? '<span class="required">*</span>' : '' ) .
              '<input id="author" class="form-control" name="author" type="text" value="' . esc_attr( $commenter['comment_author'] ) .
              '" size="30" /></p>',
    
            'email' =>
              '<p class="comment-form-email"><label for="email">' . __( 'Email', 'mn_flow' ) . '</label> ' .
              ( $req ? '<span class="required">*</span>' : '' ) .
              '<input id="email" class="form-control" name="email" type="text" value="' . esc_attr(  $commenter['comment_author_email'] ) .
              '" size="30" /></p>',
    
            'url' =>
              '<p class="comment-form-url"><label for="url">' .
              __( 'Website', 'mn_flow' ) . '</label>' .
              '<input id="url" class="form-control" name="url" type="text" value="' . esc_attr( $commenter['comment_author_url'] ) .
              '" size="30" /></p>'
            )
          ),
        );
    
        comment_form( $args ); */ ?>
    
    </div>
@endunless
`

**Comment by LeRondPoint** [comment]
Ah wait a minute...  I had left my @include('partials.comments') in there....

**Comment by LeRondPoint** [comment]
So I don't understand why your code works.  have_comments returns false for me if I haven't called comments_template before it...

**Comment by jlambe** [comment]
I think your `@if(have_comments())` has to be inside your `content.scout.php`.
So your `comments.scout.php` file only contains this code:
```php
<ol class="comment-list">
                {{
                    wp_list_comments( array(
                        'style'      => 'ol',
                        'short_ping' => true,
                        'avatar_size'=> 34,
                    ))
                }}
            </ol><!-- .comment-list -->
```

All the rest should be on `content.scout.php` because the loop statement is over there and when scout process your `comments.scout.php` it knows nothing about the loop. Let me how it goes.

**Comment by LeRondPoint** [comment]
Well I ended up with teh following: 

```
                    {{ comments_template('/resources/views/comments.php') }}
                    @include('partials.comments')
```

comments.php being empty...  lol

**Comment by jlambe** [comment]
That's weird. Theoretically, you should check if there are comments inside the single post loop. If there are comments, use the `comments_template()` function. So the layout and nav for your comments have to be inside the post loop inside your if statement for comments and not in the file passed to the comments_template function.

The file passed to `comments_template()` function only contains the code for the comments list and the list pagination.

**Comment by LeRondPoint** [comment]
Well WP documentation stats otherwise...

`Warning: this function will always return "false" until after comments_template has been called. If you need to check for comments before calling comments_template, use get_comments_number instead.`

**Comment by LeRondPoint** [comment]
In version 1.2.3, I added the following to ScoutCompiler.php.  I used a little trick to prevent having to use an empty file!!!

```
    
    /**
     * Compile the commentloop statement into valid PHP.
     *
     * @param string $file
     * @return string
     */
    protected function compileCommentloop($file)
    {
        $result = '<?php ob_start(); comments_template({$file}); ob_clean(); ?>';
        $result .= $this->compileInclude($file);
        
        return $result;
    }
```

**Comment by jlambe** [comment]
Well it's true that in my code I do have a call on `comments_number()` before using `have_comments()`😄

**Comment by LeRondPoint** [comment]
Oh wait that snippet doesn't work... I forget I need t update a theme file for it to be recompiled....

**Comment by jlambe** [comment]
Here my full code sample that works on a project:
```php
@if(comments_open())
    <section id="comments" class="comments">
        <div class="content-center">
            <h2>{{ comments_number(__('No comment.', THEME_TEXTDOMAIN)) }}</h2>
            @if(have_comments())
                {{ comments_template('/resources/views/components/comments.php') }}
            @endif
            <?php comment_form([
                'title_reply' => __('Add a comment', THEME_TEXTDOMAIN),
                'class_submit' => 'submit button button-secondary'
            ]); ?>
        </div>
    </section>
@endif
```

**Comment by LeRondPoint** [comment]
Well this works for version 1.2.3.  Since this code was removed in the master branch, I'm not going to make it cleaner and it might not work for all cases (like not using a scout template).  I haven't found the 'location' conversion code so this is it...

Anyways, I think you get the idea.  I'm not fan of the commentloop directive either but it was forced by the compiler in 1.2.3 (compileComments being taken by... comments!)

So... adapting this code to the current master branch under the @comment directive would be great. and I might get to it.  It would force me to update to master branch, are there any breaking changes I should expect?

```
    /**
     * Compile the comments statement into valid PHP.
     *
     * @param string|WP_Query $expression
     * @return string
     */
    protected function compileCommentloop($file)
    {
        if (starts_with($file, '('))
        {
            $file = substr($file, 1, -1);
        }
        
        $filePHP = str_replace("'", "", str_replace(".", "/", $file));
        $filePHP = '/resources/views/' . $filePHP . ".scout.php";
        
        $result = '<?php ob_start(); comments_template("' . $filePHP . '"); ob_end_clean(); ?>';
        $result .= $this->compileInclude($file);
        
        return $result;
    }
```

On a side note... I'd like to have a chat with you Julien.  Can you reach out at sgravel@lerondpoint.net ?

**Comment by jlambe** [comment]
The next release has a lot of refactoring but methods are the same. It has changed behind the scene only. The master branch has Scout template for backwards compatibility but the full Blade template is in place so we can more effectively add new directives like `@comment` for example. (Did I also mentioned that Twig is available in master 😄 ...)

There are break changes for next release but as always I'll give a detailed upgrade guide in order to switch from 1.2.3 to 1.3.0.

I'll reach you on monday. Regards,

**Comment by LeRondPoint** [comment]
Yeah, I tried updating to master in composer but evrything stopped working...  I didn't checkout what was going on, there's no need in the short term.

Anyways if you feel like adding the comment directive, be my guest!  Maybe commenttemplate would be a better name, sticking to WP names...  even when they suck!

One way or another, I started exploring Themosis this week and I must say it's really cool. I feel this is the direction Automattic needs to go for the next major release of WP.  It is lagging behind a lot in terms of good dev practices!  So I'm converting this theme to a more modern tech using Themosis and AngularJS, I'll share it with you when it's done.
